File: lua/problem_17.lua
-- longestFilePath returns the character length of the longest absolute
-- path to a file in the encoded filesystem string.
-- Uses a depth-indexed table of cumulative path lengths.
-- This runs in O(n) time.
local function longestFilePath(input)
    local stack = { [0] = 0 }
    local best = 0

    for line in (input .. "\n"):gmatch("([^\n]*)\n") do
        local depth = 0
        while line:sub(depth + 1, depth + 1) == "\t" do depth = depth + 1 end
        local name = line:sub(depth + 1)
        local pathLen = stack[depth] + #name
        stack[depth + 1] = pathLen + 1  -- +1 for '/'

        if name:find("%.") then  -- it's a file
            if pathLen > best then best = pathLen end
        end
    end
    return best
end

local function problem17()
    print("Problem 17:")
    local fs = "dir\n\tsubdir1\n\t\tfile1.ext\n\t\tsubsubdir1\n\tsubdir2\n\t\tsubsubdir2\n\t\t\tfile2.ext"
    assert(longestFilePath(fs) == 32, "assertion failed")
    assert(longestFilePath("dir\n\tsubdir1\n\tsubdir2\n\t\tfile.ext") == 20, "assertion failed")
    assert(longestFilePath("a") == 0, "assertion failed")
    print("All Tests Passed")
end

return { problem17 = problem17 }

File: c/problem_17.c
#include "problem_17.h"
#include <stdio.h>
#include <string.h>
#include <assert.h>

#define MAX_DEPTH 32

/* longestFilePath: depth-stack approach. O(n) time. */
int longestFilePath(const char *input) {
    int stack[MAX_DEPTH + 1];
    memset(stack, 0, sizeof(stack));
    int best = 0;
    const char *p = input;

    while (*p) {
        /* count leading tabs = depth */
        int depth = 0;
        while (*p == '\t') { depth++; p++; }
        /* read name until newline or end */
        const char *nameStart = p;
        int nameLen = 0;
        int isFile  = 0;
        while (*p && *p != '\n') {
            if (*p == '.') isFile = 1;
            nameLen++; p++;
        }
        int pathLen = stack[depth] + nameLen;
        stack[depth + 1] = pathLen + 1; /* +1 for '/' */
        if (isFile && pathLen > best) best = pathLen;
        if (*p == '\n') p++;
    }
    return best;
}

void problem17(void) {
    printf("Problem 17:\n");
    const char *fs = "dir\n\tsubdir1\n\t\tfile1.ext\n\t\tsubsubdir1\n\tsubdir2\n\t\tsubsubdir2\n\t\t\tfile2.ext";
    assert(longestFilePath(fs) == 32);
    assert(longestFilePath("dir\n\tsubdir1\n\tsubdir2\n\t\tfile.ext") == 20);
    assert(longestFilePath("a") == 0);
    printf("All Tests Passed\n");
}
